### PR TITLE
feat: add authentication support

### PR DESCRIPTION
**Pull request by rodmatos**
This pull request includes the integration between OpenAPI's `security` functionality and a client provided middleware that is executed on the routes that were marked with `security`.

Please refer to the docs/README.md for any usage questions.

### CODE REVIEW

File: example/src/AuthenticationMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { SlushyAuthenticationMiddleware, SlushyResponse, SlushyRequest } from '@slushy/server';
+
+export class AuthenticationMiddleware implements SlushyAuthenticationMiddleware {
+    execute(req: SlushyRequest, res: SlushyResponse, next: (error?: Error) => any): any {
+        if (req.headers.authorization) {
+            next()
+        } else {
+            res.status(401).json({ error: { message: "unauthorized" }})
```
Comment (aimed) [inline example/src/AuthenticationMiddleware.ts]: Ideally the library user doesn't want to do this, but instead generate an error e.g.
```ts
new ServerUnauthorizedError({ typedPayload: '' })
```
that contains the possible error codes and response bodies defined in the swagger schema.
Can we showcase this somehow? Maybe we need to extend some of the type generation things for that? 
I think the only response that is not in the swagger schema and still allowed is 500 with an empty response body.

File: server/src/Slushy.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ export class Slushy {
     public constructor(
         public readonly props: Readonly<SlushyProps>,
         public readonly app: SlushyApplication = SlushyApplicationFactory.create(),
-        public readonly router: SlushyRouter = new SlushyRouter(props, app)
```
Comment (aimed) [inline server/src/Slushy.ts]: Do we modify it now? Probably an oversight :)

File: server/src/Slushy.ts [hunk 3]
```diff
@@ -25,6 +25,7 @@ export class Slushy {
         const openApi = JSON.parse((await fs.readFile(config.resourceConfiguration.getOpenApiPath())).toString())
         const slushy = new Slushy({
             openApi,
+            authenticationMiddleware: config.authenticationMiddleware,
```
Comment (aimed) [inline server/src/Slushy.ts]: Mh this is necessary? Silly ts 🤮

Comment (rodmatos) [reply server/src/Slushy.ts]: It is not, will remove it!

**Review by aimed — DISMISSED** [review]
🥇 niceee!

File: server/src/Slushy.ts (hunk 2 above)
Comment (rodmatos) [reply server/src/Slushy.ts]: Found out another way to get it done. If you look inside `RegisterAuthenticationMiddleware` you'll that I create a new `express.Router`. That way I just register a `router` in a `router. Xzibit would be proud of me.

Comment (aimed) [reply server/src/Slushy.ts]: Haha nice, hacking express 🥇

File: example/src/AuthenticationMiddleware.ts (hunk 1 above)
Comment (aimed) [reply example/src/AuthenticationMiddleware.ts]: image

Comment (aimed) [reply example/src/AuthenticationMiddleware.ts]: I think that's the part we actually need to implement for this, so let's do it the next time! :)

Comment (rodmatos) [reply example/src/AuthenticationMiddleware.ts]: Yeah. We have to generate those types if declared in OpenAPI file. Will work on that!

Comment (aimed) [reply example/src/AuthenticationMiddleware.ts]: FYI these types are generated now :)

**Comment by rodmatos** [comment]
@aimed this is appearing in my pull requests for a while. decided that it was the time to ask: shall we do this or can I close it? are you actively supporting the project?

**Comment by aimed** [comment]
@rodmatos AHAHAH I just found it there too, this can be closed 🤗

**Pull request closed without merging**